RFCs/FS-1092-Erased-Union-Types.md

```diff
@@ -0,0 +1,253 @@
+﻿# F# RFC FS-1092 - Erased Union Types
+
+This RFC covers the detailed proposal for this suggestion. [Erased type-tagged anonymous union types](https://github.com/fsharp/fslang-suggestions/issues/538).
+
+* [ ] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/538)
+* [ ] Details: TBD
+* [ ] Implementation: [Preliminary Prototype](https://github.com/dotnet/fsharp/pull/10566)
+
+
+# Summary
+[summary]: #summary
+
+Add erased union types as a feature to F#. Erased Union types provide some of the benefits of structural ("duck") typing, within the confines of a nominative type system.
+
+# Motivation
+[motivation]: #motivation
+
+Supporting erased union types in the language allows us to move more type information with the usual advantages this brings:
+
+* They serve as an alternative to function overloading.
+* They obey subtyping rules.
+* They allow representing subset of protocols as a type without needing to resort to the lowest common denominator like `obj`.
+* Types are actually enforced, so mistakes can be caught early.
+* They allow representing more than one type
+* Because they are enforced, type information is less likely to become outdated or miss edge-cases.
+* Types are checked during inheritance, enforcing the Liskov Substitution Principle.
+
+```fsharp
+let distance(x: (Point|Location), y: (Point|Location)) = ...
+```
+
+```fsharp
+type RunWork = RunWork of args: string
+type RequestProgressUpdate = RequestProgressUpdate of workId: int
+type SubscribeProgressUpdate = SubscribeProgressUpdate of receiver: string
+type WorkerMessage = (RunWork | RequestProgressUpdate)
+type WorkManagerMessage = (RunWork | SubscribeProgressUpdate)
+
+let processWorkerMessage (msg: WorkerMessage) =
+    match msg with
+    | :? RunWork as m -> ...
+    | :? RequestProgressUpdate m -> ...
+```
+
+```fsharp
+type Username = Username of string
+type Password = Password of string
+type UserOrPass = (Password | UserName) // UserOrPass is a type alias
+
+// `getUserOrPass` is inferred to `unit -> UserOrPass`
+let getUserOrPass () = if (true) then name :> UserOrPass else password :> UserOrPass
+
+// `getUserOrPass2` binding is inferred to `unit -> (UserOrPass | Error)`
+let getUserOrPass2 () = if isErr() then err :> (UserOrPass | Error) else getUserOrPass() :> _
+```
+
+The definition of operators for types becomes simpler.
+
+```fsharp
+type Decision =
+    // Fields
+    
+    abstract member (*) (a: float, b: Decision) : LinearExpression =
+        // member body
+    abstract member (*) (a: Decision, b: Decision) : LinearExpression =
+        // member body
+```
+
+Becomes
+
+``` fsharp
+type Decision =
+    // Fields
+    
+    abstract member (*) (a: (float|Decision), b:Decision) : LinearExpression =
+        match a with
+        | :> float as f -> // float action
```

Comment:
why not `:?` instead of `:>`

Reply:
No reason. That is the more correct thing to use in this situation. I was referencing some other code that I had done which relied on casting. I would prefer `:?`.

Reply:
The big thing for me is easing reasoning and long term support of domains with A) many operator overloads and B) a more robust form of method/function overloading. Two major pain points I would much rather have this than function overloading. I believe it is a cleaner and more sustainable approach.